Fix version extraction for APsystems

alfwro13 opened the pull request:
Fixed index out of range error when inverter version number is returned empty

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
This fixes the index out of range error message when the device version number is missing - [Issue #137552 ](https://github.com/home-assistant/core/issues/137552)


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #137552 
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

alfwro13 committed:
Update entity.py

Fixed index out of range error when inverter version number is returned empty

mawoka-myblock commented:
Wouldn't it be easier to use the get method on the array?

mawoka-myblock commented:
@home-assistant rename Fix version extraction for APsystems

ronansalmon commented:
I can confirm that this fix works for me. Thanks.

frenck commented on homeassistant/components/apsystems/entity.py:
```diff
@@ -19,10 +19,19 @@ def __init__(
+        version_parts = (
+            data.coordinator.device_version.split(" ")
+            if data.coordinator.device_version
+            else ["Unknown"]
```
We shouldn't return an version if it is unknown, in that case we should set `sw_version` to `None`.

../Frenck

alfwro13 replied:
change made as per your suggestion

alfwro13 committed:
Fix: Set sw_version to None if device version is unknown

Updated logic to ensure sw_version is set to None instead of returning "Unknown" when the device version is unavailable, as per review feedback.

alfwro13 committed:
Merge branch 'dev' into dev

mawoka-myblock commented on homeassistant/components/apsystems/entity.py:
```diff
@@ -19,10 +19,19 @@ def __init__(
+        version_parts = (
+            data.coordinator.device_version.split(" ")
+            if data.coordinator.device_version
+            else ["None"]
```
```suggestion
            else [None]
```

alfwro13 committed:
Update homeassistant/components/apsystems/entity.py

Co-authored-by: Marlon <mawol@protonmail.com>

alfwro13 committed:
Merge branch 'dev' into dev

mawoka-myblock commented:
@alfwro13 Please don't do merges here. Follow this guide instead: https://developers.home-assistant.io/docs/development_catching_up/

alfwro13 committed:
Merge branch 'home-assistant:dev' into dev

frenck commented on homeassistant/components/apsystems/entity.py:
```diff
@@ -19,10 +19,19 @@ def __init__(
         """Initialize the APsystems entity."""
+
+        # Handle device version safely
+        version_parts = (
+            data.coordinator.device_version.split(" ")
+            if data.coordinator.device_version
+            else [None]
+        )
+        sw_version = version_parts[1] if len(version_parts) > 1 else version_parts[0]
```
This seems like odd handling, let's simplify it:

Something like:

```
sw_version = None
if data.coordinator.device_version:
	...
```

alfwro13 replied:
done

alfwro13 committed:
Reformated for better visualization

alfwro13 committed:
Merge pull request #1 from alfwro13/apsystems-version-fix

Reformated for better visualization

alfwro13 committed:
Merge branch 'home-assistant:dev' into dev

alfwro13 commented on homeassistant/components/apsystems/entity.py:
```diff
@@ -19,10 +19,20 @@ def __init__(
+            version_parts = data.coordinator.device_version.split(" ")
+            if len(version_parts) > 1:
+                sw_version = version_parts[1]
+            else:
+                sw_version = version_parts[0]
```
I think it should be left there to cover future versions.  

On my hardware:  
- Some inverters have `""` (empty value) → so `None`.  
- Others return `"EZ1 1.7.0"` → and we want `[1]`, so `[0]` could technically be ignored.  

So, we could simplify it to:  

```python
sw_version = None
if data.coordinator.device_version:
    version_parts = data.coordinator.device_version.split(" ")
    sw_version = version_parts[1]
```

However, I can’t say with 100% certainty that all inverters (especially on different firmware versions) will always include the `"EZ1"` prefix. If some return only a single-part version (e.g., `"1.8.0"` instead of `"EZ1 1.8.0"`), dropping the else would cause an IndexError.

So, it’s probably safer to keep the else clause to prevent potential issues in the future.

frenck reviewed (approved):
Thanks, @alfwro13 👍 

../Frenck

mawoka-myblock commented:
You see, you manged it @alfwro13!
